### PR TITLE
SONARHTML-111 Support precise issue location

### PR DESCRIPTION
**Pull request by saberduck**

### CODE REVIEW

File: sonar-html-plugin/src/main/java/org/sonar/plugins/html/core/HtmlSensor.java [hunk 1]
```diff
@@ -133,18 +133,25 @@ private static void saveMetrics(SensorContext context, HtmlSourceCode sourceCode
...
     }
   }
 
+  private static NewIssueLocation locationForIssue(InputFile inputFile, HtmlIssue issue, NewIssue newIssue) {
+    NewIssueLocation location = newIssue.newLocation()
+      .on(inputFile)
+      .message(issue.message());
+    Integer line = issue.line();
+    if (issue.startColumn() != null) {
+      location.at(inputFile.newRange(issue.line(), issue.startColumn(), issue.endLine(), issue.endColumn()));
```
Comment (quentin-jaquier-sonarsource) [inline sonar-html-plugin/src/main/java/org/sonar/plugins/html/core/HtmlSensor.java]: Do you assume that if startColumn is not null, endLine and endColumn will not be either? Or is it fine to have two last as null? 

If it's the first case I think we should change it!

File: sonar-html-plugin/src/main/java/org/sonar/plugins/html/checks/sonar/FieldsetWithoutLegendCheck.java [hunk 2]
```diff
@@ -40,12 +40,12 @@ public void startElement(TagNode node) {
   @Override
   public void endElement(TagNode node) {
     if (isFieldSet(node)) {
-      if (!foundLegend && fieldsetLine != 0) {
-        createViolation(fieldsetLine, "Add a <legend> tag to this fieldset.");
+      if (!foundLegend && fieldset != null) {
```
Comment (quentin-jaquier-sonarsource) [inline sonar-html-plugin/src/main/java/org/sonar/plugins/html/checks/sonar/FieldsetWithoutLegendCheck.java]: Is it possible that node.getStartLinePosition() returns 0?
If no, everything is fine, if yes we miss the case where fieldset is not null but equals to zero!

The same happen in FlashUsesBothObjectAndEmbedCheck.java.

Comment (saberduck) [reply sonar-html-plugin/src/main/java/org/sonar/plugins/html/checks/sonar/FieldsetWithoutLegendCheck.java]: theoretically, you are right, but the rule doesn't care about the special case of fieldset at line 0, it should detect the fieldset regardless of its position.

File: sonar-html-plugin/src/main/java/org/sonar/plugins/html/checks/HtmlIssue.java [hunk 3]
```diff
@@ -60,4 +68,16 @@ public Double cost() {
     return cost;
   }
 
+  public Integer endLine() {
```
[the remark this answers is not in the document]
Comment (saberduck) [reply sonar-html-plugin/src/main/java/org/sonar/plugins/html/checks/HtmlIssue.java]: Indeed, this was not very nice, I handled it differently, with a subclass which doesn't use null values

File: sonar-html-plugin/src/main/java/org/sonar/plugins/html/core/HtmlSensor.java (hunk 1 above)
Comment (saberduck) [reply sonar-html-plugin/src/main/java/org/sonar/plugins/html/core/HtmlSensor.java]: indeed, this different cases didn't make much sense, I think it's nicer with a specialized subclass for precise issue location

File: sonar-html-plugin/src/main/java/org/sonar/plugins/html/core/HtmlSensor.java [hunk 4]
```diff
@@ -133,18 +134,29 @@ private static void saveMetrics(SensorContext context, HtmlSourceCode sourceCode
...
       newIssue.save();
     }
   }
 
+  private static NewIssueLocation locationForIssue(InputFile inputFile, HtmlIssue issue, NewIssue newIssue) {
+    NewIssueLocation location = newIssue.newLocation()
+      .on(inputFile)
+      .message(issue.message());
+    Integer line = issue.line();
+    if (issue instanceof PreciseHtmlIssue) {
```
Comment (quentin-jaquier-sonarsource) [inline sonar-html-plugin/src/main/java/org/sonar/plugins/html/core/HtmlSensor.java]: 👍

**Pull request merged**